File: packages/native-plugins/llama/android/src/main/java/ai/eliza/plugins/llama/LlamaPlugin.kt
package ai.eliza.plugins.llama

import android.app.ActivityManager
import android.content.Context
import android.os.Build
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancelChildren
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * Capacitor bridge for llama.cpp on Android.
 *
 * The real bridge requires:
 *   1. llama.cpp sources under `android/src/main/cpp/`
 *   2. A `CMakeLists.txt` pointing at them with GGML + NEON + OpenBLAS enabled
 *   3. A JNI wrapper (e.g. `android/src/main/cpp/jni_bridge.cpp`) exposing
 *      `loadModel`, `generate`, `unloadModel` as `extern "C"` functions
 *   4. `System.loadLibrary("llama_jni")` in this class's companion `init`
 *
 * None of those exist yet — see README.md. Until then, every method returns
 * an "unavailable" error so the JS layer falls back to the server-side engine.
 */
@CapacitorPlugin(name = "Llama")
class LlamaPlugin : Plugin() {

    private val inferenceScope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
    private val inferenceMutex = Mutex()

    @Volatile
    private var loadedModelPath: String? = null

    @Volatile
    private var cancelRequested = false

    // ── Hardware probe ──────────────────────────────────────────────────

    @PluginMethod
    fun getHardwareInfo(call: PluginCall) {
        val context = context ?: run {
            call.reject("No context available")
            return
        }
        val activityManager = context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
        val memInfo = ActivityManager.MemoryInfo()
        activityManager.getMemoryInfo(memInfo)

        val totalRamGb = memInfo.totalMem / 1_073_741_824.0
        val availableRamGb = memInfo.availMem / 1_073_741_824.0
        val cpuCores = Runtime.getRuntime().availableProcessors()

        val result = JSObject().apply {
            put("platform", "android")
            put("deviceModel", "${Build.MANUFACTURER} ${Build.MODEL}")
            put("totalRamGb", totalRamGb)
            put("availableRamGb", availableRamGb)
            put("cpuCores", cpuCores)
            put("gpuSupported", NATIVE_BACKEND_AVAILABLE)
            val gpu = JSObject().apply {
                put("backend", "vulkan")
                put("available", NATIVE_BACKEND_AVAILABLE)
            }
            put("gpu", if (NATIVE_BACKEND_AVAILABLE) gpu else JSObject.NULL)
        }
        call.resolve(result)
    }

    // ── Load / unload ───────────────────────────────────────────────────

    @PluginMethod
    fun isLoaded(call: PluginCall) {
        val result = JSObject().apply {
            put("loaded", loadedModelPath != null)
            put("modelPath", loadedModelPath ?: JSObject.NULL)
        }
        call.resolve(result)
    }

    @PluginMethod
    fun loadModel(call: PluginCall) {
        if (!NATIVE_BACKEND_AVAILABLE) {
            call.reject(UNAVAILABLE_MESSAGE)
            return
        }
        val modelPath = call.getString("modelPath")
        if (modelPath.isNullOrEmpty()) {
            call.reject("modelPath is required")
            return
        }
        val contextSize = call.getInt("contextSize") ?: 4096
        val useGpu = call.getBoolean("useGpu") ?: true

        inferenceScope.launch {
            inferenceMutex.withLock {
                // TODO(llama.cpp): replace with real JNI loader:
                //   1. System.loadLibrary("llama_jni") once in a companion init
                //   2. nativeLoadModel(modelPath, contextSize, useGpu)
                //   3. throw a descriptive exception on failure
                loadedModelPath = modelPath
                call.resolve()
                // Silence unused-variable warnings until the JNI path lands.
                @Suppress("UNUSED_EXPRESSION") contextSize
                @Suppress("UNUSED_EXPRESSION") useGpu
            }
        }
    }

    @PluginMethod
    fun unloadModel(call: PluginCall) {
        inferenceScope.launch {
            inferenceMutex.withLock {
                // TODO(llama.cpp): nativeUnloadModel()
                loadedModelPath = null
                call.resolve()
            }
        }
    }

    // ── Generate ────────────────────────────────────────────────────────

    @PluginMethod
    fun generate(call: PluginCall) {
        if (!NATIVE_BACKEND_AVAILABLE) {
            call.reject(UNAVAILABLE_MESSAGE)
            return
        }
        if (loadedModelPath == null) {
            call.reject("No model loaded. Call loadModel first.")
            return
        }
        val prompt = call.getString("prompt") ?: run {
            call.reject("prompt is required")
            return
        }
        val maxTokens = call.getInt("maxTokens") ?: 512
        val temperature = call.getFloat("temperature") ?: 0.7f
        val topP = call.getFloat("topP") ?: 0.9f
        val stream = call.getBoolean("stream") ?: false

        cancelRequested = false
        val startedAt = System.currentTimeMillis()

        inferenceScope.launch {
            inferenceMutex.withLock {
                // TODO(llama.cpp): real JNI sampling loop.
                //   - tokenise prompt
                //   - decode step by step
                //   - respect cancelRequested / stopSequences
                //   - when `stream`, call notifyListeners("token", {...})
                val placeholder = "[llama.cpp native build not yet linked]"

                val duration = System.currentTimeMillis() - startedAt
                val result = JSObject().apply {
                    put("text", if (stream) "" else placeholder)
                    put("promptTokens", prompt.length / 4)
                    put("outputTokens", placeholder.length / 4)
                    put("durationMs", duration)
                }
                if (stream) {
                    notifyListeners("generationComplete", result)
                }
                call.resolve(result)

                @Suppress("UNUSED_EXPRESSION") maxTokens
                @Suppress("UNUSED_EXPRESSION") temperature
                @Suppress("UNUSED_EXPRESSION") topP
            }
        }
    }

    @PluginMethod
    fun cancelGenerate(call: PluginCall) {
        cancelRequested = true
        inferenceScope.coroutineContext[Job]?.cancelChildren()
        call.resolve()
    }

    override fun handleOnDestroy() {
        super.handleOnDestroy()
        inferenceScope.coroutineContext[Job]?.cancelChildren()
    }

    companion object {
        /**
         * Flips to true once the JNI wrapper is compiled and System.loadLibrary
         * succeeds. See README for build instructions.
         */
        private const val NATIVE_BACKEND_AVAILABLE = false
        private const val UNAVAILABLE_MESSAGE =
            "llama.cpp JNI bridge is not compiled into this build. See @elizaos/capacitor-llama/README.md for build instructions."
    }
}

File: packages/native-plugins/llama/ios/Sources/LlamaPlugin/LlamaPlugin.swift
import Capacitor
import Foundation

// MARK: - llama.cpp bridge
//
// This plugin wraps llama.cpp as an iOS `xcframework`. The framework is NOT
// committed to the repo — see README.md for the build-from-source steps.
// Until the framework is dropped into `ios/Frameworks/llama.xcframework`,
// every method returns an "unavailable" error so the JS layer falls back
// to the server-side engine without crashing.
//
// The bridge itself is intentionally a thin shim: it owns a single
// `llama_context` per plugin instance, serialises generate calls, and
// fans token streams out as Capacitor notifications.

// swiftlint:disable file_length

@objc(LlamaPlugin)
public class LlamaPlugin: CAPPlugin {
    public static let identifier = "LlamaPlugin"
    public static let jsName = "Llama"

    public override class func pluginMethods() -> [CAPPluginMethod] {
        [
            CAPPluginMethod(name: "getHardwareInfo", returnType: CAPPluginReturnPromise),
            CAPPluginMethod(name: "isLoaded", returnType: CAPPluginReturnPromise),
            CAPPluginMethod(name: "loadModel", returnType: CAPPluginReturnPromise),
            CAPPluginMethod(name: "unloadModel", returnType: CAPPluginReturnPromise),
            CAPPluginMethod(name: "generate", returnType: CAPPluginReturnPromise),
            CAPPluginMethod(name: "cancelGenerate", returnType: CAPPluginReturnPromise),
        ]
    }

    // MARK: - State

    /// Serialised queue so generate calls don't interleave on the native side.
    private let inferenceQueue = DispatchQueue(
        label: "ai.eliza.llama.inference",
        qos: .userInitiated
    )

    private var loadedModelPath: String?
    private var cancelRequested = false

    // MARK: - Hardware probing

    @objc func getHardwareInfo(_ call: CAPPluginCall) {
        let device = UIDevice.current
        let physicalMemory = ProcessInfo.processInfo.physicalMemory
        let ramGb = Double(physicalMemory) / 1_073_741_824.0
        let cpuCores = ProcessInfo.processInfo.activeProcessorCount

        call.resolve([
            "platform": "ios",
            "deviceModel": device.model,
            "totalRamGb": ramGb,
            "availableRamGb": NSNull(),
            "gpu": [
                "backend": "metal",
                "available": true,
            ],
            "cpuCores": cpuCores,
            "gpuSupported": Self.llamaBackendAvailable,
        ])
    }

    // MARK: - Load / unload

    @objc func isLoaded(_ call: CAPPluginCall) {
        call.resolve([
            "loaded": loadedModelPath != nil,
            "modelPath": loadedModelPath ?? NSNull(),
        ])
    }

    @objc func loadModel(_ call: CAPPluginCall) {
        guard Self.llamaBackendAvailable else {
            call.reject(Self.unavailableMessage)
            return
        }
        guard let modelPath = call.getString("modelPath"), !modelPath.isEmpty else {
            call.reject("modelPath is required")
            return
        }

        let contextSize = call.getInt("contextSize") ?? 4096
        let useGpu = call.getBool("useGpu") ?? true

        inferenceQueue.async { [weak self] in
            guard let self = self else { return }
            // TODO(llama.cpp): when `llama.xcframework` is integrated:
            //   1. `llama_backend_init()` once at plugin init
            //   2. `llama_model_params = llama_model_default_params()`
            //      with `n_gpu_layers = useGpu ? Int32.max : 0`
            //   3. `llama_load_model_from_file(path, params)`
            //   4. `llama_context_params.n_ctx = contextSize`
            //   5. `llama_new_context_with_model(model, ctxParams)`
            // For now, store the path and resolve so higher-level coordination
            // flows can be tested on the JS side end-to-end.
            self.loadedModelPath = modelPath
            DispatchQueue.main.async {
                call.resolve()
            }
            _ = contextSize
            _ = useGpu
        }
    }

    @objc func unloadModel(_ call: CAPPluginCall) {
        inferenceQueue.async { [weak self] in
            guard let self = self else { return }
            // TODO(llama.cpp): dispose context + model:
            //   llama_free(ctx); llama_free_model(model);
            self.loadedModelPath = nil
            DispatchQueue.main.async {
                call.resolve()
            }
        }
    }

    // MARK: - Generate

    @objc func generate(_ call: CAPPluginCall) {
        guard Self.llamaBackendAvailable else {
            call.reject(Self.unavailableMessage)
            return
        }
        guard loadedModelPath != nil else {
            call.reject("No model loaded. Call loadModel first.")
            return
        }
        guard let prompt = call.getString("prompt") else {
            call.reject("prompt is required")
            return
        }

        let maxTokens = call.getInt("maxTokens") ?? 512
        let temperature = call.getFloat("temperature") ?? 0.7
        let topP = call.getFloat("topP") ?? 0.9
        let stream = call.getBool("stream") ?? false
        let stopSequences = call.getArray("stopSequences", String.self) ?? []

        cancelRequested = false
        let startedAt = Date()
        let owner = self

        inferenceQueue.async {
            // TODO(llama.cpp): the real implementation:
            //   1. Tokenise prompt via `llama_tokenize`
            //   2. Feed tokens into the decoder via `llama_decode` batches
            //   3. On each new token, apply sampling (temperature, top_p)
            //      via `llama_sampler_chain_add`
            //   4. If `stream`, notify listeners via `notifyListeners("token", ...)`
            //   5. Check `cancelRequested` between tokens for abort support
            //   6. Honour stopSequences by scanning the cumulative string
            //
            // Until the framework is in place we surface a deterministic
            // placeholder so the higher UI can still be exercised in
            // development builds.
            let placeholder = "[llama.cpp not yet compiled for this build]"

            DispatchQueue.main.async {
                let duration = Int(Date().timeIntervalSince(startedAt) * 1000)
                let result: [String: Any] = [
                    "text": stream ? "" : placeholder,
                    "promptTokens": prompt.count / 4,
                    "outputTokens": placeholder.count / 4,
                    "durationMs": duration,
                ]
                if stream {
                    owner.notifyListeners("generationComplete", data: result)
                }
                call.resolve(result)
            }
            _ = maxTokens
            _ = temperature
            _ = topP
            _ = stopSequences
        }
    }

    @objc func cancelGenerate(_ call: CAPPluginCall) {
        cancelRequested = true
        call.resolve()
    }

    // MARK: - Backend availability

    /// Flips to true once `llama.xcframework` is linked and
    /// `llama_backend_init()` succeeds. See README for build steps.
    private static let llamaBackendAvailable = false
    private static let unavailableMessage =
        "llama.cpp framework is not compiled into this build. See @elizaos/capacitor-llama/README.md for build instructions."
}
